libcloudforensics/gcp_test.py

```diff
@@ -16,28 +16,343 @@
+
+  @mock.patch('libcloudforensics.gcp.GoogleCloudProject.gce_operation')
+  @mock.patch('libcloudforensics.gcp.GoogleCloudProject.gce_api')
+  def test_list_instances(self, mock_gce_api, mock_gce_operation):
+    """Test that instances of project are correctly listed. """
+    mock_gce_api.return_value.instances.return_value.aggregatedList.return_value.execute.return_value = None
+    mock_gce_operation.return_value = {
+      # See https://cloud.google.com/compute/docs/reference/rest/v1/instances/aggregatedList
+      # for complete structure
+      'items': {
+        0: {
+          'instances': [
+            {
+              'name': FAKE_INSTANCE.name,
+              'zone': '/' + FAKE_INSTANCE.zone
+            }
+          ]
+        }
+      }
+    }
+    instances = FAKE_ANALYSIS_PROJECT.list_instances()
+    self.assertEqual(len(instances), 1)
+    self.assertTrue(FAKE_INSTANCE.name in instances)
+    self.assertEqual(instances[FAKE_INSTANCE.name]['zone'], FAKE_INSTANCE.zone)
+
+  @mock.patch('libcloudforensics.gcp.GoogleCloudProject.gce_operation')
+  @mock.patch('libcloudforensics.gcp.GoogleCloudProject.gce_api')
+  def test_list_disks(self, mock_gce_api, mock_gce_operation):
+    """Test that disks of instances are correctly listed. """
+    mock_gce_api.return_value.disks.return_value.aggregatedList.return_value.execute.return_value = None
+    mock_gce_operation.return_value = {
+      # See https://cloud.google.com/compute/docs/reference/rest/v1/disks/aggregatedList
+      # for complete structure
+      'items': {
+        0: {
+          'disks': [
+            {
+              'name': FAKE_BOOT_DISK.name,
+              'zone': '/' + FAKE_BOOT_DISK.zone
+            }
+          ]
+        },
+        1: {
+          'disks': [
+            {
+              'name': FAKE_DISK.name,
+              'zone': '/' + FAKE_DISK.zone
+            }
+          ]
+        }
+      }
+    }
+    disks = FAKE_ANALYSIS_PROJECT.list_disks()
+    self.assertEqual(len(disks), 2)
+    self.assertTrue(FAKE_DISK.name in disks and FAKE_BOOT_DISK.name in disks)
+    self.assertEqual(disks[FAKE_DISK.name]['zone'], FAKE_DISK.zone)
+    self.assertEqual(disks[FAKE_BOOT_DISK.name]['zone'], FAKE_BOOT_DISK.zone)
+
+  @mock.patch('libcloudforensics.gcp.GoogleCloudProject.list_instances')
+  def test_get_instance(self, mock_list_instances):
+    """Test that an instance of a project can be found. """
+    mock_list_instances.return_value = {
+      FAKE_INSTANCE.name: {
+        'zone': FAKE_INSTANCE.zone
+      }
+    }
+    found_instance = FAKE_ANALYSIS_PROJECT.get_instance(FAKE_INSTANCE.name, FAKE_INSTANCE.zone)
+    self.assertIsInstance(found_instance, gcp.GoogleComputeInstance)
+    self.assertEqual(found_instance.project, FAKE_INSTANCE.project)
+    self.assertEqual(found_instance.name, FAKE_INSTANCE.name)
+    self.assertEqual(found_instance.zone, FAKE_INSTANCE.zone)
+    self.assertEqual(found_instance._data, FAKE_INSTANCE._data)
```

Comment:
As discussed offline, if you know the value of `FAKE_INSTANCE.name` in advance, it's better to actually have the string literal here rather than a variable. If it's interesting to know which variable it's supposed to be equivalent to, you can add a comment.

Reply:
Addressed, where possible